doc/design/tensor_design.md

```diff
@@ -0,0 +1,175 @@
+# Tensor Design
+
+
+## Allocation
+
+`Allocation` is a [RAII](http://en.cppreference.com/w/cpp/language/raii) class template, which is used to handle a piece of memory. 
+
+```cpp
+// Template parametr 'Device' can be 'CpuDevice' or 'GpuDevice'
```

Comment:
I also dwell on it. Maybe the key is how we name the template parameter, `Place` or `Device`?